content/courses/advanced-ada/chapters/types.rst

```diff
@@ -660,6 +660,155 @@ for :ada:`A_Plus`.
 We'll see later how definite and indefinite types apply to
 :ref:`formal parameters <Formal_Definite_Indefinite_Subtypes>`.
 
+Constrained Attribute
+~~~~~~~~~~~~~~~~~~~~~
+
+We can use the :ada:`Constrained` attribute to verify whether an object of
+discriminated type is constrained or not. Let's start our discussion by reusing
+the :ada:`Simple_Record` type from previous examples. In this version of the
+:ada:`Unconstrained_Types` package, we're adding a :ada:`Reset` procedure for
+the discriminated record type:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Types.Constrained_Attribute
+
+    package Unconstrained_Types is
+
+       type Simple_Record (Extended : Boolean := False) is record
+          V : Integer;
+          case Extended is
+             when False =>
+                null;
+             when True  =>
+                V_Float : Float;
+          end case;
+       end record;
+
+       procedure Reset (R : in out Simple_Record);
+
+    end Unconstrained_Types;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    package body Unconstrained_Types is
+
+       procedure Reset (R : in out Simple_Record) is
+          Zero_Not_Extended : constant Simple_Record :=
+            (Extended => False, V => 0);
+          Zero_Extended : constant Simple_Record :=
+            (Extended => True, V => 0, V_Float => 0.0);
+       begin
+          Put_Line ("---- Reset: R'Constrained => " & R'Constrained'Image);
+
+          if not R'Constrained then
+             R := Zero_Extended;
+          else
+             if R.Extended then
+                R := Zero_Extended;
+             else
+                R := Zero_Not_Extended;
+             end if;
+         end if;
+       end Reset;
+
+    end Unconstrained_Types;
+
+As the name indicates, the :ada:`Reset` procedure initializes all record
+components with zero. Note that we use the :ada:`Constrained` attribute to
+verify whether objects are constrained before assigning to them. For objects
+that are not constrained, we can simply assign another object to it |mdash| as
+we do with the :ada:`R := Zero_Extended` statement. When an object is
+constrained, however, the discriminants must match. If we assign an object to
+:ada:`R`, the discriminants of that object must match the discriminants of
```

Comment:
"the discriminants of that object must match the discriminants" => 
"the discriminant of that object must match the discriminant"

(since there is only one discriminant for R)